File: go/pkg/recall/segment.go
package recall

import (
	"context"
	"math"
	"sort"
	"strings"

	"github.com/vmihailenco/msgpack/v5"
)

// Score fusion weights for combining search signals.
const (
	weightVector  = 0.5
	weightKeyword = 0.3
	weightLabel   = 0.2
)

// StoreSegment persists a segment in the index.
//
// It stores the segment data in KV (msgpack-encoded, keyed by timestamp),
// and if an embedder and vector index are configured, embeds the summary
// text and inserts the resulting vector.
func (idx *Index) StoreSegment(ctx context.Context, seg Segment) error {
	data, err := msgpack.Marshal(seg)
	if err != nil {
		return err
	}

	key := segmentKey(idx.prefix, seg.Timestamp)
	if err := idx.store.Set(ctx, key, data); err != nil {
		return err
	}

	// Index the vector if both embedder and vector index are available.
	if idx.embedder != nil && idx.vec != nil {
		vec, err := idx.embedder.Embed(ctx, seg.Summary)
		if err != nil {
			return err
		}
		if err := idx.vec.Insert(seg.ID, vec); err != nil {
			return err
		}
	}

	return nil
}

// DeleteSegment removes a segment by ID.
//
// Since segment keys are time-based (not ID-based), this scans all
// segments to find the matching ID. For typical per-persona indexes
// (hundreds to low thousands of segments) this is acceptable.
func (idx *Index) DeleteSegment(ctx context.Context, id string) error {
	prefix := segmentPrefix(idx.prefix)

	for entry, err := range idx.store.List(ctx, prefix) {
		if err != nil {
			return err
		}
		var seg Segment
		if err := msgpack.Unmarshal(entry.Value, &seg); err != nil {
			continue // skip malformed entries
		}
		if seg.ID == id {
			if err := idx.store.Delete(ctx, entry.Key); err != nil {
				return err
			}
			// Remove from vector index if configured.
			if idx.vec != nil {
				_ = idx.vec.Delete(id)
			}
			return nil
		}
	}

	return nil // not found is not an error
}

// GetSegment retrieves a segment by ID.
//
// Scans all segments to find the matching ID. Returns nil if not found.
func (idx *Index) GetSegment(ctx context.Context, id string) (*Segment, error) {
	prefix := segmentPrefix(idx.prefix)

	for entry, err := range idx.store.List(ctx, prefix) {
		if err != nil {
			return nil, err
		}
		var seg Segment
		if err := msgpack.Unmarshal(entry.Value, &seg); err != nil {
			continue
		}
		if seg.ID == id {
			return &seg, nil
		}
	}

	return nil, nil
}

// RecentSegments returns the n most recent segments, ordered newest first.
//
// KV keys are lexicographically ordered by date+timestamp, so we scan
// all segments and take the last n.
func (idx *Index) RecentSegments(ctx context.Context, n int) ([]Segment, error) {
	if n <= 0 {
		return nil, nil
	}

	prefix := segmentPrefix(idx.prefix)
	var all []Segment

	for entry, err := range idx.store.List(ctx, prefix) {
		if err != nil {
			return nil, err
		}
		var seg Segment
		if err := msgpack.Unmarshal(entry.Value, &seg); err != nil {
			continue
		}
		all = append(all, seg)
	}

	// KV list is ascending; reverse to get newest first.
	for i, j := 0, len(all)-1; i < j; i, j = i+1, j-1 {
		all[i], all[j] = all[j], all[i]
	}

	if len(all) > n {
		all = all[:n]
	}
	return all, nil
}

// SearchSegments searches for segments matching the query using multi-signal
// scoring: vector similarity, keyword overlap, and label overlap.
//
// The scoring pipeline:
//  1. Collect all segments (with optional time filtering)
//  2. If embedder + vec are available: embed query text → vector search → distance scores
//  3. Keyword score: fraction of query terms found in segment keywords
//  4. Label score: fraction of segment labels overlapping with query labels
//  5. Fuse: 0.5*vector + 0.3*keyword + 0.2*label
//  6. Sort by score descending, return top Limit
func (idx *Index) SearchSegments(ctx context.Context, q SearchQuery) ([]ScoredSegment, error) {
	limit := q.Limit
	if limit <= 0 {
		limit = 10
	}

	// Step 1: Load all segments with time filtering.
	segments, err := idx.loadSegments(ctx, q)
	if err != nil {
		return nil, err
	}
	if len(segments) == 0 {
		return nil, nil
	}

	// Step 2: Vector scores (ID → score in [0,1]).
	vecScores := make(map[string]float64)
	if idx.embedder != nil && idx.vec != nil && q.Text != "" && idx.vec.Len() > 0 {
		queryVec, err := idx.embedder.Embed(ctx, q.Text)
		if err != nil {
			return nil, err
		}
		// Request more candidates than limit to account for filtering.
		topK := limit * 3
		if topK < 20 {
			topK = 20
		}
		matches, err := idx.vec.Search(queryVec, topK)
		if err != nil {
			return nil, err
		}
		for _, m := range matches {
			// Convert distance to similarity score in [0,1].
			// Assuming cosine distance in [0,2], similarity = 1 - distance/2.
			sim := 1.0 - float64(m.Distance)/2.0
			if sim < 0 {
				sim = 0
			}
			vecScores[m.ID] = sim
		}
	}

	// Step 3+4+5: Score each segment and fuse signals.
	queryTerms := tokenize(q.Text)
	labelSet := toSet(q.Labels)
	hasVec := len(vecScores) > 0
	hasKeywords := len(queryTerms) > 0
	hasLabels := len(labelSet) > 0

	scored := make([]ScoredSegment, 0, len(segments))
	for _, seg := range segments {
		score := 0.0

		// Vector signal.
		if hasVec {
			if vs, ok := vecScores[seg.ID]; ok {
				score += weightVector * vs
			}
		}

		// Keyword signal: fraction of query terms found in segment keywords.
		if hasKeywords {
			score += weightKeyword * keywordScore(queryTerms, seg.Keywords)
		}

		// Label signal: fraction of segment labels in query label set.
		if hasLabels {
			score += weightLabel * labelScore(seg.Labels, labelSet)
		}

		// Skip zero-score segments unless no filters were applied.
		if score == 0 && (hasVec || hasKeywords || hasLabels) {
			continue
		}

		scored = append(scored, ScoredSegment{Segment: seg, Score: score})
	}

	// Step 6: Sort by score descending, then by timestamp descending.
	sort.Slice(scored, func(i, j int) bool {
		if scored[i].Score != scored[j].Score {
			return scored[i].Score > scored[j].Score
		}
		return scored[i].Segment.Timestamp > scored[j].Segment.Timestamp
	})

	if len(scored) > limit {
		scored = scored[:limit]
	}
	return scored, nil
}

// loadSegments scans all segments from KV, applying time and label filters.
func (idx *Index) loadSegments(ctx context.Context, q SearchQuery) ([]Segment, error) {
	prefix := segmentPrefix(idx.prefix)
	afterNs := int64(0)
	beforeNs := int64(math.MaxInt64)
	if !q.After.IsZero() {
		afterNs = q.After.UnixNano()
	}
	if !q.Before.IsZero() {
		beforeNs = q.Before.UnixNano()
	}

	labelSet := toSet(q.Labels)
	filterLabels := len(labelSet) > 0

	var segments []Segment
	for entry, err := range idx.store.List(ctx, prefix) {
		if err != nil {
			return nil, err
		}
		var seg Segment
		if err := msgpack.Unmarshal(entry.Value, &seg); err != nil {
			continue
		}

		// Time filter.
		if seg.Timestamp < afterNs || seg.Timestamp >= beforeNs {
			continue
		}

		// Label filter: segment must share at least one label.
		if filterLabels && !hasOverlap(seg.Labels, labelSet) {
			continue
		}

		segments = append(segments, seg)
	}
	return segments, nil
}

// keywordScore computes the fraction of query terms found in the segment's
// keywords. Both are compared in lowercase for case-insensitive matching.
func keywordScore(queryTerms []string, segKeywords []string) float64 {
	if len(queryTerms) == 0 {
		return 0
	}
	segSet := make(map[string]struct{}, len(segKeywords))
	for _, kw := range segKeywords {
		segSet[strings.ToLower(kw)] = struct{}{}
	}
	hits := 0
	for _, qt := range queryTerms {
		if _, ok := segSet[qt]; ok {
			hits++
		}
	}
	return float64(hits) / float64(len(queryTerms))
}

// labelScore computes the fraction of segment labels present in the query
// label set. Returns 0 if the segment has no labels.
func labelScore(segLabels []string, queryLabelSet map[string]struct{}) float64 {
	if len(segLabels) == 0 {
		return 0
	}
	hits := 0
	for _, l := range segLabels {
		if _, ok := queryLabelSet[l]; ok {
			hits++
		}
	}
	return float64(hits) / float64(len(segLabels))
}

// tokenize splits text into lowercase terms for keyword matching.
func tokenize(text string) []string {
	if text == "" {
		return nil
	}
	fields := strings.Fields(strings.ToLower(text))
	// Deduplicate.
	seen := make(map[string]struct{}, len(fields))
	result := make([]string, 0, len(fields))
	for _, f := range fields {
		if _, ok := seen[f]; !ok {
			seen[f] = struct{}{}
			result = append(result, f)
		}
	}
	return result
}

// toSet converts a string slice to a set (map).
func toSet(items []string) map[string]struct{} {
	if len(items) == 0 {
		return nil
	}
	m := make(map[string]struct{}, len(items))
	for _, item := range items {
		m[item] = struct{}{}
	}
	return m
}

// hasOverlap returns true if any element of items is in the set.
func hasOverlap(items []string, set map[string]struct{}) bool {
	for _, item := range items {
		if _, ok := set[item]; ok {
			return true
		}
	}
	return false
}

File: rust/recall/src/segment.rs
use std::collections::HashSet;

use crate::error::RecallError;
use crate::index::RecallIndex;
use crate::keys::{bucket_prefix, parse_sid_value, segment_key, segment_prefix, sid_key, sid_value};
use crate::types::{Bucket, ScoredSegment, SearchQuery, Segment};

/// Score fusion weights for combining search signals.
const WEIGHT_VECTOR: f64 = 0.5;
const WEIGHT_KEYWORD: f64 = 0.3;
const WEIGHT_LABEL: f64 = 0.2;

impl RecallIndex {
    /// Persist a segment in the index.
    ///
    /// The segment's bucket field determines which bucket it is stored in.
    /// If bucket is empty, it defaults to "1h".
    pub async fn store_segment(&self, seg: &Segment) -> Result<(), RecallError> {
        let bucket = if seg.bucket.is_empty() { Bucket::BUCKET_1H } else { &seg.bucket };

        let data = rmp_serde::to_vec_named(seg)
            .map_err(|e| RecallError::Serialization(e.to_string()))?;

        let seg_k = segment_key(&self.prefix, bucket, seg.timestamp);
        let sid_k = sid_key(&self.prefix, &seg.id);
        let sid_v = sid_value(bucket, seg.timestamp);

        let old_loc = self.lookup_segment_loc(&seg.id);

        self.store.batch_set(&[
            (seg_k.as_str(), data.as_slice()),
            (sid_k.as_str(), sid_v.as_bytes()),
        ])?;

        if let Ok((old_bucket, old_ts)) = old_loc {
            if old_ts != seg.timestamp || old_bucket != bucket {
                let old_key = segment_key(&self.prefix, &old_bucket, old_ts);
                let _ = self.store.delete(&old_key);
            }
        }

        if let (Some(embedder), Some(vec)) = (&self.embedder, &self.vec) {
            let embedding = embedder.embed(&seg.summary).await?;
            vec.insert(&seg.id, &embedding)?;
        }

        Ok(())
    }

    /// Remove a segment by ID.
    pub fn delete_segment(&self, id: &str) -> Result<(), RecallError> {
        let (bucket, ts) = match self.lookup_segment_loc(id) {
            Ok(loc) => loc,
            Err(_) => return Ok(()),
        };

        let seg_k = segment_key(&self.prefix, &bucket, ts);
        let sid_k = sid_key(&self.prefix, id);
        self.store.batch_delete(&[seg_k.as_str(), sid_k.as_str()])?;

        if let Some(vec) = &self.vec {
            let _ = vec.delete(id);
        }
        Ok(())
    }

    /// Retrieve a segment by ID. Returns None if not found.
    pub fn get_segment(&self, id: &str) -> Result<Option<Segment>, RecallError> {
        let (bucket, ts) = match self.lookup_segment_loc(id) {
            Ok(loc) => loc,
            Err(_) => return Ok(None),
        };

        let data = match self.store.get(&segment_key(&self.prefix, &bucket, ts))? {
            Some(data) => data,
            None => return Ok(None),
        };

        let seg: Segment = rmp_serde::from_slice(&data)
            .map_err(|e| RecallError::Serialization(e.to_string()))?;
        Ok(Some(seg))
    }

    /// Return the n most recent segments across all buckets, newest first.
    pub fn recent_segments(&self, n: usize) -> Result<Vec<Segment>, RecallError> {
        if n == 0 {
            return Ok(vec![]);
        }

        let prefix = segment_prefix(&self.prefix);
        let entries = self.store.scan(&prefix)?;

        let mut all = Vec::new();
        for (_key, value) in entries {
            if let Ok(seg) = rmp_serde::from_slice::<Segment>(&value) {
                all.push(seg);
            }
        }

        all.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));
        all.truncate(n);
        Ok(all)
    }

    /// Return all segments in a specific bucket in KV scan order.
    ///
    /// The underlying KV contract guarantees scan returns keys in sorted order.
    /// With segment keys encoded as `{prefix}:seg:{bucket}:{ts_20d}`, this is the
    /// effective ordering used by compaction.
    pub fn bucket_segments(&self, bucket: &Bucket) -> Result<Vec<Segment>, RecallError> {
        let prefix = bucket_prefix(&self.prefix, bucket.as_str());
        let entries = self.store.scan(&prefix)?;

        let mut segments = Vec::new();
        for (_key, value) in entries {
            if let Ok(seg) = rmp_serde::from_slice::<Segment>(&value) {
                segments.push(seg);
            }
        }
        Ok(segments)
    }

    /// Return segment count and total summary char count in a bucket.
    pub fn bucket_stats(&self, bucket: &Bucket) -> Result<(usize, usize), RecallError> {
        let prefix = bucket_prefix(&self.prefix, bucket.as_str());
        let entries = self.store.scan(&prefix)?;

        let mut count = 0;
        let mut chars = 0;
        for (_key, value) in entries {
            if let Ok(seg) = rmp_serde::from_slice::<Segment>(&value) {
                count += 1;
                chars += seg.summary.len();
            }
        }
        Ok((count, chars))
    }

    /// Search segments using multi-signal scoring.
    pub async fn search_segments(
        &self,
        q: &SearchQuery,
    ) -> Result<Vec<ScoredSegment>, RecallError> {
        let limit = if q.limit == 0 { 10 } else { q.limit };

        let segments = self.load_segments(q)?;
        if segments.is_empty() {
            return Ok(vec![]);
        }

        let mut vec_scores = std::collections::HashMap::new();
        if let (Some(embedder), Some(vec)) = (&self.embedder, &self.vec) {
            if !q.text.is_empty() && vec.len() > 0 {
                let query_vec = embedder.embed(&q.text).await?;
                let top_k = (limit * 3).max(20);
                let matches = vec.search(&query_vec, top_k)?;
                for m in matches {
                    let sim = (1.0 - m.distance as f64 / 2.0).max(0.0);
                    vec_scores.insert(m.id, sim);
                }
            }
        }

        let query_terms = tokenize(&q.text);
        let label_set: HashSet<&str> = q.labels.iter().map(|s| s.as_str()).collect();
        let has_vec = !vec_scores.is_empty();
        let has_keywords = !query_terms.is_empty();
        let has_labels = !label_set.is_empty();

        let mut scored = Vec::new();
        for seg in segments {
            let mut score = 0.0;

            if has_vec {
                if let Some(&vs) = vec_scores.get(&seg.id) {
                    score += WEIGHT_VECTOR * vs;
                }
            }

            if has_keywords {
                score += WEIGHT_KEYWORD * keyword_score(&query_terms, &seg.keywords);
            }

            if has_labels {
                score += WEIGHT_LABEL * label_score(&seg.labels, &label_set);
            }

            if score == 0.0 && (has_vec || has_keywords || has_labels) {
                continue;
            }

            scored.push(ScoredSegment {
                segment: seg,
                score,
            });
        }

        scored.sort_by(|a, b| {
            b.score
                .partial_cmp(&a.score)
                .unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| b.segment.timestamp.cmp(&a.segment.timestamp))
        });

        scored.truncate(limit);
        Ok(scored)
    }

    fn load_segments(&self, q: &SearchQuery) -> Result<Vec<Segment>, RecallError> {
        let prefix = segment_prefix(&self.prefix);
        let entries = self.store.scan(&prefix)?;

        let after_ns = q.after_ns;
        let before_ns = if q.before_ns == 0 { i64::MAX } else { q.before_ns };
        let label_set: HashSet<&str> = q.labels.iter().map(|s| s.as_str()).collect();
        let filter_labels = !label_set.is_empty();

        let mut segments = Vec::new();
        for (_key, value) in entries {
            let seg: Segment = match rmp_serde::from_slice(&value) {
                Ok(s) => s,
                Err(_) => continue,
            };

            if seg.timestamp < after_ns || seg.timestamp >= before_ns {
                continue;
            }

            if filter_labels && !seg.labels.iter().any(|l| label_set.contains(l.as_str())) {
                continue;
            }

            segments.push(seg);
        }
        Ok(segments)
    }

    fn lookup_segment_loc(&self, id: &str) -> Result<(String, i64), RecallError> {
        let key = sid_key(&self.prefix, id);
        match self.store.get(&key)? {
            Some(data) => {
                parse_sid_value(&data).map_err(|e| RecallError::Serialization(e))
            }
            None => Err(RecallError::Storage("not found".into())),
        }
    }
}

fn keyword_score(query_terms: &[String], seg_keywords: &[String]) -> f64 {
    if query_terms.is_empty() {
        return 0.0;
    }
    let seg_set: HashSet<String> = seg_keywords.iter().map(|kw| kw.to_lowercase()).collect();
    let hits = query_terms.iter().filter(|qt| seg_set.contains(*qt)).count();
    hits as f64 / query_terms.len() as f64
}

fn label_score(seg_labels: &[String], query_label_set: &HashSet<&str>) -> f64 {
    if seg_labels.is_empty() {
        return 0.0;
    }
    let hits = seg_labels
        .iter()
        .filter(|l| query_label_set.contains(l.as_str()))
        .count();
    hits as f64 / seg_labels.len() as f64
}

fn tokenize(text: &str) -> Vec<String> {
    if text.is_empty() {
        return vec![];
    }
    let mut seen = HashSet::new();
    text.to_lowercase()
        .split_whitespace()
        .filter(|w| seen.insert(w.to_string()))
        .map(|w| w.to_string())
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tokenize() {
        assert_eq!(tokenize(""), Vec::<String>::new());
        assert_eq!(tokenize("Hello World"), vec!["hello", "world"]);
        assert_eq!(tokenize("hello hello world"), vec!["hello", "world"]);
        assert_eq!(tokenize("  A  B  "), vec!["a", "b"]);
    }

    #[test]
    fn test_keyword_score_fn() {
        let query = vec!["hello".to_string(), "world".to_string()];
        let kw = vec!["Hello".to_string(), "Foo".to_string()];
        let score = keyword_score(&query, &kw);
        assert!((score - 0.5).abs() < 0.001);
    }

    #[test]
    fn test_label_score_fn() {
        let labels = vec!["a".to_string(), "b".to_string(), "c".to_string()];
        let query_set: HashSet<&str> = ["a", "c"].into_iter().collect();
        let score = label_score(&labels, &query_set);
        assert!((score - 2.0 / 3.0).abs() < 0.001);
    }
}
